code/game/machinery/merch_vendor.dm

```diff
@@ -0,0 +1,124 @@
+			var/obj/item/I = merch.typepath
+			var/pp = replacetext(replacetext("[merch.typepath]", "/obj/item/", ""), "/", "-")
+			imagelist[pp] = "[icon2base64(icon(initial(I.icon), initial(I.icon_state), SOUTH, 1))]"
+
+/obj/machinery/economy/merch/attackby(obj/item/I, mob/user)
+	if(isspacecash(I))
+		insert_cash(I, user)
+
+	return ..()
+
+/obj/machinery/economy/merch/proc/do_purchase(datum/money_account/customer_account, datum/merch_item/merch, mob/user)
+	if(!merch)
+		return FALSE
+
+	if(cash_stored >= merch.cost)
+		pay_with_cash(merch.cost, "Purchase of [merch.name]", name, user, account_database.vendor_account)
+		give_change(user)
+	else
+		var/mob/living/carbon/human/H = user
+		if(istype(H))
+			var/obj/item/card/id/C = H.get_idcard(TRUE)
+			if(!C || !pay_with_card(C, merch.cost, "Purchase of [merch.name]", "NAS Trurl Merchandising", user, account_database.vendor_account))
+				return FALSE
+		else
+			to_chat(user, "<span class='warning'>Payment failure: you have no ID or other method of payment.")
+			return FALSE
+
+	deliver(merch, user)
+	return TRUE
+
+/obj/machinery/economy/merch/proc/deliver(datum/merch_item/item, mob/user)
+	var/obj/item/merch = new item.typepath(get_turf(src))
+	var/obj/item/smallDelivery/D = new(get_turf(src))
+	D.name = "small parcel - 'Your Nanotrasen Swag'"
+	D.wrapped = merch
+	merch.forceMove(D)
+	user.put_in_hands(D)
+	SSeconomy.total_vendor_transactions++
+
+/obj/machinery/economy/merch/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.physical_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "MerchVendor", name, 450, 500, master_ui, state)
+		ui.open()
+
+/obj/machinery/economy/merch/ui_data(mob/user)
+	var/list/data = list()
+
+	data["inserted_cash"] = cash_transaction
+	var/datum/money_account/user_account = user.get_worn_id_account()
+	data["user_cash"] = user_account?.credit_balance
+
+	return data
+
+/obj/machinery/economy/merch/ui_static_data(mob/user)
+	var/list/static_data = list()
+
+	static_data["imagelist"] = imagelist
+	static_data["products"] = list()
+	for(var/category in merch_categories)
+		static_data["products"][category] = list()
+		for(var/datum/merch_item/merch in merchandise[category])
+			var/list/product_data = list(
+				"name" = merch.name,
+				"price" = merch.cost,
+				"desc" = merch.desc,
+				"category" = merch.category,
+				"path" = replacetext(replacetext("[merch.typepath]", "/obj/item/", ""), "/", "-"),
+			)
+			static_data["products"][category] += list(product_data)
+
+	return static_data
+
+/obj/machinery/economy/merch/ui_act(action, params, datum/tgui/ui)
+	if(..())
+		return
+	if(stat & (NOPOWER|BROKEN))
+		return
+
+	. = FALSE
+	var/mob/living/user = ui.user
+
+	switch(action)
+		if("purchase")
+			. = TRUE
+			for(var/datum/merch_item/merch in merchandise[params["category"]])
+				if(merch.name == params["name"])
```

Comment:
Can there be more matches than just one? If not then please add a break at the end of the logic

Reply:
added